db/job.go

```diff
@@ -913,3 +657,109 @@ func scanJobs(conn Conn, lockFactory lock.LockFactory, rows *sql.Rows) (Jobs, er
 
 	return jobs, nil
 }
+
+func (j *job) SyncResourceSpaceCombinations(combinations []map[string]string) ([]JobCombination, error) {
+	tx, err := j.conn.Begin()
+	if err != nil {
+		return nil, err
+	}
+
+	defer Rollback(tx)
+
+	jobCombinations := []JobCombination{}
+
+	for _, c := range combinations {
+		var jobCombinationID int
+
+		marshaled, err := json.Marshal(c)
+		if err != nil {
+			return nil, err
+		}
+
+		needsMigration, err := findInvalidJobCombination(tx, j.ID())
+		if err != nil {
+			return nil, err
+		}
+
+		if needsMigration {
+			err = psql.Update("job_combinations").
+				Set("combination", marshaled).
+				Where(sq.Eq{"job_id": j.ID()}).
+				Suffix("RETURNING id").
+				RunWith(tx).QueryRow().Scan(&jobCombinationID)
+			if err != nil {
+				return nil, err
+			}
+		} else {
+			err = psql.Insert("job_combinations").
+				Columns("job_id", "combination").
+				Values(j.ID(), marshaled).
+				// FIXME: postgres requires an update to return id on conflict, but this may cause side effects
+				Suffix("ON CONFLICT (job_id, combination) DO UPDATE set combination=EXCLUDED.combination RETURNING id").
```

Comment:
Rather than doing a write here, how about we `DO NOTHING` and then just select the ID after? That will prevent continuous writes from occurring in a stable state.